### PR TITLE
kubeadm: don't use the Docker SDK in util/system/docker_validator*

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:

Instead of creating a Docker client and fetching an Info object
from the Docker enpoint, call the "docker info" command
and populate a local dockerInfo struct from JSON output.

Also add unit tests.

This change affects "test/e2e_node/e2e_node_suite_test.go"
as it consumes this Docker validator by calling
"system.ValidateSpec()".

this decouples the validator and kubeadm from the Docker SDK.

**Which issue(s) this PR fixes**:

xref https://github.com/kubernetes/kubeadm/issues/1638

**Special notes for your reviewer**:
NONE

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```
should not be user facing.

/assign @yastij 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
cc @dims

/kind cleanup
/priority important-soon

### CODE REVIEW

**Comment by neolit123** [comment]
/sig node
/cc @Random-Liu
as this affects https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/e2e_node_suite_test.go
xref https://github.com/kubernetes/kubernetes/pull/32427

File: cmd/kubeadm/app/util/system/docker_validator.go [hunk 1]
```diff
@@ -51,18 +58,28 @@ func (d *DockerValidator) Validate(spec SysSpec) (error, error) {
 		return nil, nil
 	}
 
-	c, err := client.NewClient(dockerEndpoint, "", nil, nil)
+	// Run 'docker info' with a JSON output and unmarshal it into a dockerInfo object
+	info := dockerInfo{}
+	out, err := exec.Command("docker", "info", "--format", "{{json .}}").Output()
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to create docker client")
+		return nil, errors.Wrap(err, `failed executing "docker info --format '{{json .}}'"`)
```
Comment (andrewsykim) [inline cmd/kubeadm/app/util/system/docker_validator.go]: Probably worth while to put the output in the error, maybe also use `CombinedOutput` to capture stderr above?

Comment (neolit123) [reply cmd/kubeadm/app/util/system/docker_validator.go]: yes, makes sense. i've added it.

File: cmd/kubeadm/.import-restrictions [hunk 2]
```diff
@@ -112,15 +112,7 @@
 				"github.com/coreos/etcd/pkg/tlsutil",
 				"github.com/coreos/etcd/pkg/transport",
 				"github.com/davecgh/go-spew/spew",
-				"github.com/docker/distribution/digestset",
 				"github.com/docker/distribution/reference",
```
Comment (neolit123) [inline cmd/kubeadm/.import-restrictions]: looks like we depend on docker still; possibly due to the ipvs imports [1]

File: cmd/kubeadm/.import-restrictions [hunk 3]
```diff
@@ -112,15 +112,7 @@
...
-				"github.com/docker/docker/api",
-				"github.com/docker/docker/client",
-				"github.com/docker/docker/pkg/term",
-				"github.com/docker/go-connections/nat",
-				"github.com/docker/go-connections/sockets",
-				"github.com/docker/go-connections/tlsconfig",
-				"github.com/docker/go-units",
 				"github.com/docker/libnetwork/ipvs",
```
Comment (neolit123) [inline cmd/kubeadm/.import-restrictions]: [1]

File: cmd/kubeadm/app/util/system/docker_validator.go [hunk 4]
```diff
@@ -32,6 +31,14 @@ type DockerValidator struct {
 	Reporter Reporter
 }
 
+// dockerInfo holds a local subset of the Info struct from
+// github.com/docker/docker/api/types.
+// The JSON output from 'docker info' should map to this struct.
+type dockerInfo struct {
+	Driver        string
```
Comment (timothysc) [inline cmd/kubeadm/app/util/system/docker_validator.go]: shouldn't the struct contain the `json: "FieldName"` annotation

Comment (dixudx) [reply cmd/kubeadm/app/util/system/docker_validator.go]: Right. Should have the json tags. There will be deserializations.

File: cmd/kubeadm/.import-restrictions (hunk 3 above)
Comment (yastij) [reply cmd/kubeadm/.import-restrictions]: This one is mostly pulling consts IIRC, I'll send a PR to consolidate our consts (rbac, this one and some others)

File: cmd/kubeadm/app/util/system/docker_validator.go (hunk 4 above)
Comment (yastij) [reply cmd/kubeadm/app/util/system/docker_validator.go]: +1

File: cmd/kubeadm/.import-restrictions (hunk 3 above)
Comment (yastij) [reply cmd/kubeadm/.import-restrictions]: This is mostly pulled to use consts IIRC, I'll send a PR to consolidate our const usage (rbac, kube-proxy etc..)

File: cmd/kubeadm/app/util/system/docker_validator.go (hunk 4 above)
Comment (neolit123) [reply cmd/kubeadm/app/util/system/docker_validator.go]: technically, it is not needed. the json deserializer in stdlib is happy if the struct field to match the name of the json field and it's not case sensitive. also we don't have serialization here.

but i will add the tags.

**Comment by neolit123** [comment]
/hold cancel
@timothysc @yastij @dixudx 
updated to include json tags.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 

thx @neolit123

**Review by yastij — COMMENTED** [review]
/lgtm

**Pull request merged**